test/itkOMEZarrNGFFInMemoryTest.cxx

```diff
@@ -72,22 +82,51 @@ doTest(const char * inputFileName, const char * outputFileName)
   {
     itkGenericExceptionMacro("The image read through memory is different from the one read through file");
   }
-  buffer = std::vector<char>(); // deallocate it
 
-  bufferInfo = itk::OMEZarrNGFFImageIO::BufferInfo{};
+  // Verify a local copy of the buffer is maintained in the itk::Image
+  memImage->DisconnectPipeline(); // keep our local copy of the memory buffer
+  std::fill(buffer.begin(), buffer.end(), 0);   // reset the memory buffer in place
+  ITK_TRY_EXPECT_NO_EXCEPTION(comparer->Update()); // should not reflect buffer update
+  if (comparer->GetNumberOfPixelsWithDifferences() > 0)
+  {
+    itkGenericExceptionMacro("After deallocating the memory buffer, the image read through memory is different from "
+                             "the one read through file");
+  }
+
+  // Verify our memory buffer is still valid
+  ITK_TEST_EXPECT_EQUAL(bufferInfo.pointer, buffer.data());
+  ITK_TEST_EXPECT_EQUAL(bufferInfo.size, buffer.size());
+
+  // Write image back to in-memory buffer as a zip compressed bitstream
   using WriterType = itk::ImageFileWriter<ImageType>;
   typename WriterType::Pointer writer = WriterType::New();
   writer->SetInput(image);
   writer->SetFileName(memAddress);
   writer->SetImageIO(zarrIO);
   ITK_TRY_EXPECT_NO_EXCEPTION(writer->Update());
 
-  // Write zip bitstream to disk
+  // Verify the output buffer occupies a new memory region with the expected size
+  ITK_TEST_EXPECT_EQUAL(bufferInfo.size, inputBufferSize);
+  ITK_TEST_EXPECT_TRUE(bufferInfo.pointer != inputBufferPointer);
+  std::cout << "Input: " << static_cast<void *>(inputBufferPointer) << " " << inputBufferSize << std::endl;
+  std::cout << "Input mem Address: " << memAddress << std::endl;
+  std::cout << "Output: " << static_cast<void *>(bufferInfo.pointer) << " " << bufferInfo.size << std::endl;
+  std::cout << "Output mem address: " << itk::OMEZarrNGFFImageIO::MakeMemoryFileName(bufferInfo) << std::endl;
```

Comment:
@dzenanz I am wondering if this behavior is expected. Could you confirm whether we expect a "write" operation to an in-memory block to allocate a new buffer for the output?

I see the following printouts when running locally:

```
8: Input: 00000199115D1E80 47685
8: Input mem Address: 366326051184.memory
8: Output: 0000019911609290 47685
8: Output mem address: 366326051184.memory
```

Observed behavior is as follows:
1. The original in-memory buffer is itself at address `00000199115D1E80` and is described by a `BufferInfo` instance that has address `366326051184`
2. _A new output buffer is allocated on write_ at address `0000019911609290`. The `BufferInfo` instance at `366326051184` is updated during the write operation to describe the new address, losing the reference to the original in-memory buffer in the process.

It seems that the content of both the input and output buffers match our expectations, as we are able to subsequently write the zip stream to file and then read it back in directly as an ITK image. If this behavior is as expected I will keep this check here. If it is instead expected that the original buffer should be overwritten then we should enter an issue to track.

Reply:
On write, it is expected that a new memory will be allocated. It is generally not known in advance how much memory will be needed, and we let minizip's internals to have growing memory buffer for storage.

If the old buffer needs deallocation, it should be done by the invoking code.